05.lsディレクトリにls.rbファイルを追加いたしました

masamune17 opened the pull request:
05.lsディレクトリにls.rbファイルを追加いたしました

masamune17 committed:
05.lsディレクトリにls.rbファイルを追加いたしました

masamune17 committed:
line49に'total'の追加、line50にlongformat情報をまとめた配列を追加、line54にlongformatの出力メソッドの追加

maedana commented on 05.ls/ls.rb:
```diff
@@ -0,0 +1,80 @@
+  end
+end
+
+file_entries = Dir.glob('*', File::FNM_DOTMATCH).sort if option.include?('a')
```
最初のほうでやってる
```ruby
file_entries = Dir.glob('*').sort
```
も、この辺でやったほうがわかりやすいかなーと思いました

maedana commented on 05.ls/ls.rb:
```diff
@@ -0,0 +1,80 @@
+end
+
+if file_entries.size == 4
+  list_segments_output_exception(file_entries)
```
この分岐はできれば無くしたいところですね。

Rubyの配列には
https://docs.ruby-lang.org/ja/latest/method/Array/i/transpose.html
という便利メソッドがあるので、これを使うとこのへんもっとシンプルなロジックに出来そうかなーと思います。いかがでしょうか?

ちなみに要件は満たしてますし、早く先に進みたいのであればこのままでも絶対に駄目とまでは言わないです。

maedana commented on 05.ls/ls.rb:
```diff
@@ -0,0 +1,80 @@
+def list_segments_output_longformat(file_entries)
+  blocks, permissions, links, users, groups, file_sizes, times, paths = Array.new(8).map { [] }
+  hash1 = { '0' => '---', '1' => '--x', '2' => '-w-', '3' => '-wx', '4' => 'r--', '5' => 'r-x', '6' => 'rw-', '7' => 'rwx' }
+  hash2 = { '00' => '-', '40' => 'd', '20' => 'l' }
```
ローカル変数とはいえちょっと命名が乱暴すぎるかなと感じます。他のものはしっかり命名しているので、ここだけ浮いているように見えますね

maedana commented on 05.ls/ls.rb:
```diff
@@ -0,0 +1,80 @@
+    printf '% -*s', max_group_length, "  #{longformats[3][i]}"
+    printf '% *s', max_file_size_length, "  #{longformats[4][i]}"
+    printf '%*s', 13, "#{longformats[5][i]} "
+    puts format (longformats[6][i]).to_s
```
0,1,2,3,4,5,6が何を示すのかパッとわからないですよね。ここはリーダブルにしたいなーと思いました。具体的には0から6それぞれを定数にしてしまって人間が見てわかるようにするか、
```ruby
longformats = [permissions, links, users, groups, file_sizes, times, paths]
```
を配列ではなくhashにしてキー名でわかるようにするのがよさそうかなと思いました。

あと引数をたくさん渡してますが、それら全てlongformatsから導出可能ですよね? なので引数はlongformatsだけにしたほうがわかりやすいかなと思いました。

このあたりいかがでしょうか

masamune17 committed:
line9~15での宣言をline58~64に移動

masamune17 committed:
line28,29

masamune17 committed:
line28,29の変数名を変更、outputメソッドへの引数をハッシュへの変更と引数のメソッド内での導出に変更

masamune17 committed:
line74~80の分岐を削除、list_segments_outputメソッドへの引数を1つにし、Araay.transposeメソッドを使用して処理を簡略化

masamune17 committed:
line74~80の分岐を削除、list_segments_outputメソッドへの引数を1つにし、transposeメソッドを使用した処理に変更

masamune17 committed:
Merge branch 'ls' of https://github.com/Masamune-git/ruby-practices into ls

maedana commented on 05.ls/ls.rb:
```diff
@@ -32,47 +27,51 @@ def list_segments_output_exception(file_entries)
-  hash2 = { '00' => '-', '40' => 'd', '20' => 'l' }
+  blocks, permissions, links, users, groups, file_sizes, times, paths = Array.new(8).map { [] }
+  change1 = { '0' => '---', '1' => '--x', '2' => '-w-', '3' => '-wx', '4' => 'r--', '5' => 'r-x', '6' => 'rw-', '7' => 'rwx' }
+  change2 = { '00' => '-', '40' => 'd', '20' => 'l' }
```
hashからchangeに変わってますが、変更の意図がよくわからないですね。なぜchangeなんでしょう? permissionを組み立てるために使う変換データだからでしょうか。だとしてもchange1, change2というほど同じものを指しているわけでもないと思いますし、短すぎて意味がよくわからない命名に思えます。他には思いつかない感じでしょうか

細かいことに感じるかもですが、他がちゃんとしてるだけにやっぱり気になりますね

masamune17 committed:
変数名の変更

line30 (変更前)change1 (変更後)convert_to_permission
line31 (変更前)change2 (変更後)convert_to_filetype
